### PR TITLE
Glue: Validate target query in correlations page

### LINKED ISSUE #55625: Glue: Validate if target query is valid
_Background_

An idea that @gelicia suggested is: Add a button to validate the target query.  Why? In https://github.com/grafana/grafana/issues/55369 a target query editor was added. However, there's no way to test if the query is valid or if it returns any results.

In later stages, we plan to display the results. However, by doing simple validation now we could add a simpler initial implementation that would hugely improve UX with smaller effort. It will also help us work on running Correlations Page queries earlier and use the same logic to run the source query. 

_Requirements_

* Add a "Test query" button (can be a different text) near the target query editor
* When a button is clicked the query is run
* A message below the query is shown below the target query
  * an error if no data frames are returned
  * info that the query works fine otherwise 

Note: The validation may stop properly working in Stage 2. In stage 2 we will allow using variables in the query that correspond to the field name, but there will be no sample results available to fill in the variables (sample results will be added in Stage 3) so queries will run not interpolated and may fail.

The benefit of doing it at this stage is that we will figure out how to run queries on Correlations Page, and it will be easier to
implement Source Editor in Stage 3.

_Technical details_

The correlations page can retrieve the corresponding data source API and run `.query` method to get results (like in [Explore](https://github.com/grafana/grafana/blob/802d67eecaecb129308b0b3d966c73ec01358ca6/public/app/features/query/state/runRequest.ts#L113), it may require using some default time-range or time range selector)

### PR DESCRIPTION
**Pull request by L-M-K-B**
Fixes #55625

### CODE REVIEW

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 1]
```diff
@@ -25,6 +49,28 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
+      runner.run({
+        datasource: datasource,
+        queries: [value],
+        timezone: 'utc',
+        timeRange: getDefaultTimeRange(),
+        maxDataPoints: 100,
+        minInterval: null,
+      });
+    }
+    await runner.get().subscribe((panelData) => {
```
Comment (L-M-K-B) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:63]: await isn't necessary as Observables are "entirely synchronous, just like a function" (see [here](https://rxjs.dev/guide/observable))

File: public/app/features/correlations/CorrelationsPage.test.tsx [hunk 2]
```diff
@@ -218,6 +229,25 @@ const renderWithContext = async (
...
+    destroy: () => {},
+  });
+
+  setQueryRunnerFactory(factory);
+
+  return emitter;
+};
+
+let emit: { subscriber: Subscriber<PanelData> | undefined } = { subscriber: undefined };
+
```
Comment (Elfo404) [inline public/app/features/correlations/CorrelationsPage.test.tsx:250]: what about:
```suggestion
type NextFn = (value?: PanelData) => void;
const mockQueryRunner = () => {
  let next: NextFn = () => {};
  const observable: Observable<PanelData> = new Observable((subscriber) => {
    next = subscriber.next;
  });
  const factory: QueryRunnerFactory = () => ({
    get: () => observable,
    run: () => {},
    cancel: () => {},
    destroy: () => {},
  });

  setQueryRunnerFactory(factory);

  return next;
};

let emit: NextFn = () => {};
```

and use it like

```ts
emit({ series: [], timeRange: getDefaultTimeRange(), state: LoadingState.Error });
```

so we don't have to do ugly non-null assertion? 😛

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 3]
```diff
@@ -13,6 +24,19 @@ interface Props {
...
+  const runner = useMemo(() => createQueryRunner(), []);
+
+  const styles = useMemo(() => {
+    return {
+      valid: css`
+        color: ${theme.colors.success.text};
+      `,
+    };
+  }, [theme]);
+
```
Comment (Elfo404) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:39]: what about using `usingStyles2` instead? something like

```ts
function getValidStyle(theme: GrafanaTheme2) {
  return {
    valid: css`
      color: ${theme.colors.success.text};
    `,
  };
}

export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
  const [isValidQuery, setIsValidQuery] = useState<boolean | undefined>(undefined);
  const styles = useStyles2(getValidStyle);
  // ...
```

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 4]
```diff
@@ -13,6 +24,19 @@ interface Props {
 }
 
 export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
+  const [isValidQuery, setIsValidQuery] = useState<boolean | undefined>(undefined);
+  const theme = useTheme2();
+
+  const runner = useMemo(() => createQueryRunner(), []);
```
Comment (Elfo404) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:30]: nit:
```suggestion
  const runner = useMemo(createQueryRunner, []);
```

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 5]
```diff
@@ -25,6 +49,29 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
+        if (!panelData || panelData.state === 'Error') {
+          setIsValidQuery(false);
+        } else if (panelData.state === 'Done') {
+          setIsValidQuery(true);
+        } else {
+          setIsValidQuery(undefined);
+        }
+      });
+    }
+  };
```
Comment (Elfo404) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:73]: we don't need `handleValidation` to be marked as async (`subscribe` does not return a promise/we don't need to use its return value) 
```suggestion
  const handleValidation = (value: DataQuery) => {
    if (datasource) {
      runner.run({
        datasource: datasource,
        queries: [value],
        timezone: 'utc',
        timeRange: getDefaultTimeRange(),
        maxDataPoints: 100,
        minInterval: null,
      });

      runner.get().subscribe((panelData) => {
        if (!panelData || panelData.state === 'Error') {
          setIsValidQuery(false);
        } else if (panelData.state === 'Done') {
          setIsValidQuery(true);
        } else {
          setIsValidQuery(undefined);
        }
      });
    }
  };
```

Comment (Elfo404) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:73]: also (this is more of a nit than anything) we could wrap it in a useCallback so it doesn't get recreated on every render (
```suggestion
  const handleValidation = async (value: DataQuery) => {
    if (datasource) {
      runner.run({
        datasource: datasource,
        queries: [value],
        timezone: 'utc',
        timeRange: getDefaultTimeRange(),
        maxDataPoints: 100,
        minInterval: null,
      });

      await runner.get().subscribe((panelData) => {
        if (!panelData || panelData.state === 'Error') {
          setIsValidQuery(false);
        } else if (panelData.state === 'Done') {
          setIsValidQuery(true);
        } else {
          setIsValidQuery(undefined);
        }
      });
    }
  };
```

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 6]
```diff
@@ -13,6 +24,19 @@ interface Props {
 }
 
 export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
+  const [isValidQuery, setIsValidQuery] = useState<boolean | undefined>(undefined);
```
Comment (Elfo404) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:27]: wouldn't `isQueryValid` be a better name instead?

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 7]
```diff
@@ -52,8 +99,33 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
+              <HorizontalGroup justify="flex-end">
+                {isValidQuery ? (
+                  <div className={styles.valid}>
+                    <Icon name="check" /> This query is valid.
+                  </div>
+                ) : isValidQuery === false ? (
+                  <FieldValidationMessage>This query is not valid.</FieldValidationMessage>
+                ) : (
+                  <div />
+                )}
```
Comment (Elfo404) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:122]: do we really need to render that empty div?
```suggestion
                ) : null}
```

**Review by Elfo404 — DISMISSED** [review]
great stuff, seems to work fine! 🚀  just a few comments over code style, lmk what you think

File: public/app/features/correlations/CorrelationsPage.test.tsx (hunk 2 above)
Comment (L-M-K-B) [reply public/app/features/correlations/CorrelationsPage.test.tsx:250]: I'd love to accept your suggestion @Elfo404 but somehow it suppresses the label for the validation and therefore the test fails. Do you have any idea why this happens?
Screenshot 2022-11-28 at 16 15 23

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 3 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:39]: In order to make the use of our style theme more consistent I'll move it as you suggested.

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 5 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:73]: 1. I agree and I'll change it.
2. I don't understand yet why this should get recreated with rerendering. `handleValidation()` is only called when the related button gets clicked and the QueryRunner itself is only created once as it is already wrapped in `useMemo()`

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 6 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:35]: `isQueryValid` sounds to me like it is a question. Can you please explain a bit more why it would be better in your opinion?

File: public/app/features/correlations/CorrelationsPage.test.tsx (hunk 2 above)
Comment (L-M-K-B) [reply public/app/features/correlations/CorrelationsPage.test.tsx]: Applied suggested changes from the following thread: https://raintank-corp.slack.com/archives/C02TXB92G6T/p1669656594428139

Comment (L-M-K-B) [reply public/app/features/correlations/CorrelationsPage.test.tsx:250]: Dropped the suggestion. See thread in Slack (link above)

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 5 above)
Comment (Elfo404) [reply public/app/features/correlations/Forms/QueryEditorField.tsx]: this is not about calling `handleValidation` but defining the function itself. every time the state or a prop changes the component rerenders - and by doing so recreates the `handleValidation` function. this happens every time you click on `Validate Query`.

but anyway it's not a biggie, we pass an anonymous function to the validate button anyway so it won't make much of a difference.

Comment (Elfo404) [reply public/app/features/correlations/Forms/QueryEditorField.tsx]: there's another issue with this though, by subscribing every time we run the validation (by calling `runner.get().subscribe(..)` we are basically creating a subscription every time you click on the button, which is a memory leak. if you put a console statement inside the subscriber you'll see how it runs once more every time you click.

we should probably move the subscription somewhere else and make sure to unsubscribe when needed. i think something like
```ts
  useEffect(() => {
    runner.get().subscribe((panelData) => {
      console.log(panelData);
      if (!panelData || panelData.state === 'Error') {
        setIsValidQuery(false);
      } else if (panelData.state === 'Done') {
        setIsValidQuery(true);
      } else {
        setIsValidQuery(undefined);
      }
    });

    return runner.destroy;
  }, [runner]);
```
(outside of the `handelValidation` function would suffice)

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 6 above)
Comment (Elfo404) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:35]: that's usually how those boolean variables are read imho

in `isQueryValid` being `Valid` a classifying adjective to `Query`, it completely describes the variable state, which is going to be `true` if the query is valid. "is Query valid?" is a complete sentence in english.

`isValidQuery` on the other hand lacks an adjective. "is valid Query?" is not a complete sentence in english, and it lacks and adjective to be complete as - in this form - "valid query" is the subject of the question.

the rule i usually follow is to name boolean variables and functions returning booleans as `is[subject][classifier]` and functions as `is[classifier](subject)`. i.e. `isWindowOpen` and `isOpen(window)`

in other words, what we are interested in is if the query is valid (isQueryValid), not if something is a valid query. if `isValidQuery` was a function i would have agreed with you when used like `isValidQuery(myQuery)`. 

anyway this was more a nit than anything else, feel free to ignore.

Comment (osg-grafana) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:35]: Responding here after a DM from @Elfo404 to check with a technical writer; my response was:

"`isQueryValid` says to me, "Is the | this query valid?"
"`isValidQuery` says to me, "[It] is a valid query." or "The | This query is valid."

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 5 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx]: I see why this needs to be changed. I didn't expect so many subscriptions and just changed it to what you suggested.

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 6 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:39]: I prefer to leave it as it is.

**Comment by L-M-K-B** [comment]
@Elfo404 and I talked about what is still missing:

- [ ] currently an empty query returns "query is valid" => should be invalid
- [ ] add test for behaviour with empty query
- [ ] `handleValidation()` should be done like in Explore but not reuse the functions from there as they are Explore specific
- [ ] add test => if there is no query result it shows an error
- [ ] sometimes the valid/invalid message doesn't disappear immediately when something is changed in the the query builder (depending on datasource) => onChange is handled in different ways 
=> I think I will add a bug issue for the last one in order to get this first version of the validation ready asap

**Comment by L-M-K-B** [comment]
Opened https://github.com/grafana/grafana/issues/60692 for the last bullet point of my previous comment

**Comment by L-M-K-B** [comment]
**Summary of possible cases for validation:**
- if the query does not return an error (LoadingState.Done) and there is data (series.length >= 1) => valid
- if query returns an error and there's no data => invalid
- if the query returns an error but there's data =>  invalid
- if the query does not return an error (LoadingState.Done) and there is not data (sereies.length == 0) => invalid

No feedback yet for users if the query is empty. There is an error message in the console.
Possible solutions:
1. Show the error message _This query is not valid._
2. Show another error message.
3. Disable _Validate query_ button until there is content in the query.
4. Don't show anything to the users.

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 8]
```diff
@@ -23,8 +50,53 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
+          __interval_ms: { text: intervalMs, value: intervalMs },
+        },
+      },
+      id,
+      done: false,
+    };
+
+    if (datasource) {
+      runRequest(datasource, transaction.request).subscribe((panelData) => {
+        // console.log(panelData)
```
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:84]:
```suggestion
```

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 9]
```diff
@@ -23,8 +50,53 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
+        if (!panelData || panelData.state === 'Error' || panelData.series.length === 0) {
+          setIsValidQuery(false);
```
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:86]: Probably it's okay for now but I wonder if it'd make sense to distinguish in the future "This query is not valid" from "This query does not return any results".

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 10]
```diff
@@ -23,8 +50,53 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
     return getDataSourceSrv().get(dsUid);
   }, [dsUid]);
+
   const QueryEditor = datasource?.components?.QueryEditor;
 
+  const handleValidation = (value: DataQuery) => {
+    const interval = '1s';
+    const intervalMs = 1000;
+    const id = generateKey();
+    const queries = [{ ...value, refId: 'something' }];
```
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:60]:
```suggestion
    const queries = [{ ...value, refId: 'A' }];
```

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 8 above)
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:84]:
```suggestion
```

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 11]
```diff
@@ -52,8 +124,31 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
+              />
+              <HorizontalGroup justify="flex-end">
+                {isValidQuery ? (
+                  <div className={style.valid}>
+                    <Icon name="check" /> This query is valid.
+                  </div>
+                ) : isValidQuery === false ? (
+                  <FieldValidationMessage>This query is not valid.</FieldValidationMessage>
+                ) : null}
+                <Button variant="primary" icon={'check'} type="button" onClick={() => handleValidation(value)}>
```
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:146]: Would it make sense to make it a `secondary` button? It's optional action and being close to the "Save" button looks like it is as important as saving.

Comment (Elfo404) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:146]: +1

https://internal-ops-us-east-0.grafana.net/design-system/Components/button/#donts

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 9 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:86]: I agree and suggest to open a follow up task for UX improvements of the validation which contains the following subtasks:
- distinguish "This query is not valid" from "This query does not return any results"
- feedback for users if the query is empty ( see [comment](https://github.com/grafana/grafana/pull/57245#issuecomment-1369007499) above)

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 11 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:146]: 👍

File: public/app/features/correlations/Forms/QueryEditorField.tsx [hunk 12]
```diff
@@ -23,8 +50,52 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
...
+          __interval_ms: { text: intervalMs, value: intervalMs },
+        },
+      },
+      id,
+      done: false,
+    };
+
+    if (datasource) {
+      runRequest(datasource, transaction.request).subscribe((panelData) => {
+        if (!panelData || panelData.state === 'Error' || panelData.series.length === 0) {
```
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.tsx:84]: When checking for number of series we also need to check the status - otherwise it will flash "not valid" message during Loading state.

```suggestion
if (
          !panelData ||
          panelData.state === 'Error' ||
          (panelData.state === 'Done' && panelData.series.length === 0)
        ) {
```

**Review by ifrost — CHANGES_REQUESTED** [review]

File: public/app/features/correlations/Forms/QueryEditorField.tsx (hunk 9 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.tsx:89]: Opened https://github.com/grafana/grafana/issues/60978

File: public/app/features/correlations/Forms/QueryEditorField.test.tsx [hunk 13]
```diff
@@ -31,9 +32,28 @@ const renderWithContext = async (
   setDataSourceSrv(dsServer);
 
   render(<Wrapper>{children}</Wrapper>);
+
+  return dsServer.get;
```
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.test.tsx:36]: Where is it being used?

File: public/app/features/query/state/queryAnalytics.ts [hunk 14]
```diff
@@ -71,7 +71,7 @@ export function emitDataRequestEvent(datasource: DataSourceApi) {
     eventData.totalQueries = totalQueries;
     eventData.cachedQueries = cachedQueries;
 
-    const dashboard = getDashboardSrv().getCurrent();
+    const dashboard = getDashboardSrv().dashboard;
```
Comment (ifrost) [inline public/app/features/query/state/queryAnalytics.ts:74]: Hmmm... not sure if we should change it just to make tests work. Could we add `silenceConsoleOutput` to our tests? I think it was intentional to use `getCurrent` here.

Comment (Elfo404) [reply public/app/features/query/state/queryAnalytics.ts:74]: if we pass the app to the request as well and we modify `https://github.com/grafana/grafana/blob/74167b4d44119862c1246caa3367e4f488efd15e/public/app/features/query/state/queryAnalytics.ts#L35` then there's no need for this change. this is only used for analytics, and would probably be a better solution then just silencing console output

File: public/app/features/correlations/Forms/QueryEditorField.test.tsx (hunk 13 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.test.tsx]: I checked it and removed it.

File: public/app/features/query/state/queryAnalytics.ts (hunk 14 above)
Comment (L-M-K-B) [reply public/app/features/query/state/queryAnalytics.ts]: Done.

File: public/app/features/query/state/queryAnalytics.ts [hunk 15]
```diff
@@ -32,8 +32,8 @@ export function emitDataRequestEvent(datasource: DataSourceApi) {
       duration: data.request.endTime! - data.request.startTime,
```
Comment (Elfo404) [inline public/app/features/query/state/queryAnalytics.ts:32]: I'd revert the changes in this file, i don't think we actually need query analytics for the correlations setup page, i think the important bit was to set `data.request.app` to  `CoreApp.Correlations` so that the code below is not triggered. wdyt @ifrost

Comment (Elfo404) [reply public/app/features/query/state/queryAnalytics.ts:32]: or better, something like 
```
if(explore){
  doExploreStuff();
} else if (dashboard) {
  doDashboardStuff();
}
```

Comment (L-M-K-B) [reply public/app/features/query/state/queryAnalytics.ts:32]: If we don't need the query analytics I'd go with Gio's second suggestion the first one is breaking the tests.

File: public/app/features/correlations/Forms/QueryEditorField.test.tsx [hunk 16]
```diff
@@ -63,4 +81,97 @@ describe('QueryEditorField', () => {
       await screen.findByRole('alert', { name: 'Data source does not export a query editor.' })
     ).toBeInTheDocument();
   });
+
+  describe('Query validation', () => {
+    beforeEach(() => {
+      // Request errors cause the query runner to print to console,
+      // this means the test would fail because of the fail on log rules.
+      // We mock console.error to avoid this.
+      jest.spyOn(console, 'error').mockImplementation();
```
Comment (ifrost) [inline public/app/features/correlations/Forms/QueryEditorField.test.tsx:90]: I don't see any errors if I remove this bit (or change to `jest.spyOn(console, 'error').mockImplementation(() => { throw new Error() });`) so it doesn't seem like it's being called. Can we remove it?

**Review by ifrost — APPROVED** [review]

File: public/app/features/correlations/Forms/QueryEditorField.test.tsx (hunk 16 above)
Comment (L-M-K-B) [reply public/app/features/correlations/Forms/QueryEditorField.test.tsx]: I removed it as it was leftover from the test improvements.

File: public/app/features/query/state/queryAnalytics.ts (hunk 15 above)
Comment (ifrost) [reply public/app/features/query/state/queryAnalytics.ts]: This tracking is important for Usage Insights (https://github.com/grafana/grafana/pull/59931). I think we should keep it.

Comment (L-M-K-B) [reply public/app/features/query/state/queryAnalytics.ts:32]: After consulting with Piotr I reverted the changes from the commit below.
Screenshot 2023-01-12 at 10 33 11
